core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -467,308 +472,61 @@ public Snapshot currentSnapshot() {
-        lastAssignedPartitionId, defaultSortOrderId, sortOrders, properties, currentSnapshotId, filtered,
-        ImmutableList.copyOf(newSnapshotLog), addPreviousFile(metadataFileLocation, lastUpdatedMillis));
-  }
-
-  private TableMetadata setCurrentSnapshotTo(Snapshot snapshot) {
-    ValidationException.check(snapshotsById.containsKey(snapshot.snapshotId()),
-        "Cannot set current snapshot to unknown: %s", snapshot.snapshotId());
-    ValidationException.check(formatVersion == 1 || snapshot.sequenceNumber() <= lastSequenceNumber,
-        "Last sequence number %s is less than existing snapshot sequence number %s",
-        lastSequenceNumber, snapshot.sequenceNumber());
-
-    if (currentSnapshotId == snapshot.snapshotId()) {
-      // change is a noop
-      return this;
-    }
-
-    long nowMillis = System.currentTimeMillis();
-    List<HistoryEntry> newSnapshotLog = ImmutableList.<HistoryEntry>builder()
-        .addAll(snapshotLog)
-        .add(new SnapshotLogEntry(nowMillis, snapshot.snapshotId()))
-        .build();
-
-    return new TableMetadata(null, formatVersion, uuid, location,
-        lastSequenceNumber, nowMillis, lastColumnId, currentSchemaId, schemas, defaultSpecId, specs,
-        lastAssignedPartitionId, defaultSortOrderId, sortOrders, properties, snapshot.snapshotId(), snapshots,
-        newSnapshotLog, addPreviousFile(metadataFileLocation, lastUpdatedMillis));
+    List<Snapshot> toRemove = snapshots.stream().filter(removeIf).collect(Collectors.toList());
+    return new Builder(this).removeSnapshots(toRemove).build();
   }
 
   public TableMetadata replaceProperties(Map<String, String> rawProperties) {
     ValidationException.check(rawProperties != null, "Cannot set properties to null");
     Map<String, String> newProperties = unreservedProperties(rawProperties);
-    TableMetadata metadata = new TableMetadata(null, formatVersion, uuid, location,
-        lastSequenceNumber, System.currentTimeMillis(), lastColumnId, currentSchemaId, schemas, defaultSpecId, specs,
-        lastAssignedPartitionId, defaultSortOrderId, sortOrders, newProperties, currentSnapshotId, snapshots,
-        snapshotLog, addPreviousFile(metadataFileLocation, lastUpdatedMillis, newProperties));
 
-    int newFormatVersion = PropertyUtil.propertyAsInt(rawProperties, TableProperties.FORMAT_VERSION, formatVersion);
-    if (formatVersion != newFormatVersion) {
-      metadata = metadata.upgradeToFormatVersion(newFormatVersion);
-    }
-
-    return metadata;
-  }
-
-  public TableMetadata removeSnapshotLogEntries(Set<Long> snapshotIds) {
-    List<HistoryEntry> newSnapshotLog = Lists.newArrayList();
-    for (HistoryEntry logEntry : snapshotLog) {
-      if (!snapshotIds.contains(logEntry.snapshotId())) {
-        // copy the log entries that are still valid
-        newSnapshotLog.add(logEntry);
+    Set<String> removed = Sets.newHashSet(properties.keySet());
```

Comment:
In `UpdateProperties`, we formulate a set of `updates` and `removes`, we merge them with the base metadata and pass to this method. But then we are doing the reverse here to decompose them back to `updated` and `removed`, it seems like redundant work. Is this done just to preserve the public `UpdateProperties` interface and `TableMetadata.replaceProperties` method?

Reply:
I considered changing how this works, but I decided not to alter the `replaceProperties` method. For one, I didn't want to make unnecessary changes to released `TableMetadata` methods. Removing `replaceProperties` can be done in a separate PR if we want.

A second reason is that the implementation of `UpdateProperties` is responsible for turning the requested changes into a final set of properties and it uses that final set to validate metrics configuration. I think it makes sense for that implementation to be fully responsible for applying changes and validating the actual result, without relying on assumptions about how changes are applied afterward. Another way of thinking about this is that the final set of properties should be whatever `UpdateProperties` produces, and this PR should not change that.

But, even if `UpdateProperties` is responsible for producing the final set of properties, the right way to keep track of what changed is using the properties that were updated and removed. That's because the correctness of replacing the set of properties implicitly relies on the starting state. If two writers both add a property concurrently (based on the same starting set) then you can't blindly run the two replace operations. So sending "replace properties" to a REST catalog requires either failing and letting the client retry, but sending "set properties" and "remove properties" changes allows the catalog to handle it.